File: xkcd/src/main/java/xyz/jienan/xkcd/whatif/fragment/SingleWhatIfFragment.java
package xyz.jienan.xkcd.whatif.fragment;

import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.text.Html;
import android.text.method.LinkMovementMethod;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.bumptech.glide.Glide;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.lang.ref.WeakReference;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import butterknife.BindView;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import timber.log.Timber;
import xyz.jienan.xkcd.BuildConfig;
import xyz.jienan.xkcd.R;
import xyz.jienan.xkcd.base.BaseFragment;
import xyz.jienan.xkcd.model.WhatIfModel;
import xyz.jienan.xkcd.model.persist.SharedPrefManager;
import xyz.jienan.xkcd.ui.WhatIfWebView;
import xyz.jienan.xkcd.whatif.interfaces.ImgInterface;
import xyz.jienan.xkcd.whatif.interfaces.LatexInterface;
import xyz.jienan.xkcd.whatif.interfaces.RefInterface;

import static android.view.HapticFeedbackConstants.CONTEXT_CLICK;
import static android.view.HapticFeedbackConstants.FLAG_IGNORE_GLOBAL_SETTING;
import static android.view.HapticFeedbackConstants.LONG_PRESS;
import static xyz.jienan.xkcd.Const.FIRE_GO_WHAT_IF_MENU;
import static xyz.jienan.xkcd.Const.FIRE_SHARE_BAR;
import static xyz.jienan.xkcd.Const.FIRE_WHAT_IF_IMG_LONG;
import static xyz.jienan.xkcd.Const.FIRE_WHAT_IF_REF;
import static xyz.jienan.xkcd.Const.FIRE_WHAT_IF_SUFFIX;

/**
 * Created by jienanzhang on 03/03/2018.
 */

public class SingleWhatIfFragment extends BaseFragment implements ImgInterface.ImgCallback, RefInterface.RefCallback {

    @BindView(R.id.webview_what_if)
    public WhatIfWebView webView;

    private int id = -1;

    private WhatIfMainFragment parentFragment;

    private LatexInterface latexInterface = new LatexInterface();

    protected CompositeDisposable compositeDisposable = new CompositeDisposable();

    private AlertDialog dialog = null;

    protected SharedPrefManager sharedPref = new SharedPrefManager();

    public static SingleWhatIfFragment newInstance(int articleId) {
        SingleWhatIfFragment fragment = new SingleWhatIfFragment();
        Bundle args = new Bundle();
        args.putInt("id", articleId);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    protected int getLayoutResId() {
        return R.layout.fragment_what_if_single;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Bundle args = getArguments();
        if (args != null)
            id = args.getInt("id", -1);
        setHasOptionsMenu(true);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = super.onCreateView(inflater, container, savedInstanceState);
        webView.getSettings().setBuiltInZoomControls(true);
        webView.getSettings().setUseWideViewPort(Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT);
        webView.getSettings().setJavaScriptEnabled(true);
        webView.getSettings().setDisplayZoomControls(false);
        webView.getSettings().setLoadWithOverviewMode(true);
        webView.getSettings().setAllowFileAccess(true);
        webView.getSettings().setAppCacheEnabled(true);
        webView.getSettings().setCacheMode(WebSettings.LOAD_DEFAULT);
        webView.getSettings().setTextZoom(sharedPref.getWhatIfZoom());
        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.KITKAT) {
            WebView.setWebContentsDebuggingEnabled(BuildConfig.DEBUG);
        }
        if (id != -1) {
            final WhatIfModel model = WhatIfModel.getInstance();

            compositeDisposable.add(model.loadArticle(id)
                    .doOnSuccess(model::push)
                    .subscribe(article -> {
                                if (webView != null) {
                                    webView.loadDataWithBaseURL("file:///android_asset/",
                                            article.content.replaceAll("\\$", "&#36;"), "text/html", "UTF-8", null);
                                }
                            },
                            Timber::e));

            webView.setCallback(new WebViewScrollCallback(this));
            webView.setLatexScrollInterface(latexInterface);
            webView.addJavascriptInterface(latexInterface, "AndroidLatex");
            webView.addJavascriptInterface(new ImgInterface(this), "AndroidImg");
            webView.addJavascriptInterface(new RefInterface(this), "AndroidRef");

            if (getParentFragment() instanceof WhatIfMainFragment) {
                parentFragment = ((WhatIfMainFragment) getParentFragment());
            }
        }
        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.KITKAT) {
            WebView.setWebContentsDebuggingEnabled(BuildConfig.DEBUG);
        }
        setHasOptionsMenu(true);
        compositeDisposable.add(WhatIfModel
                .getInstance()
                .observeZoom()
                .subscribe(zoom -> webView.getSettings().setTextZoom(zoom),
                        e -> Timber.e(e, "observing zoom error")));
        return view;
    }

    @Override
    public void onDestroyView() {
        if (dialog != null) {
            dialog.dismiss();
        }
        if (compositeDisposable != null) {
            compositeDisposable.dispose();
        }
        if (webView != null) {
            Timber.d("OkHttp: clear %d", id);
            webView.clearTasks();
        }
        super.onDestroyView();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_share:
                Intent shareIntent = new Intent();
                shareIntent.setAction(Intent.ACTION_SEND);
                shareIntent.putExtra(Intent.EXTRA_TEXT, getString(R.string.share_text_what_if, "https://whatif.xkcd.com/" + id));
                shareIntent.setType("text/plain");
                startActivity(Intent.createChooser(shareIntent, getResources().getText(R.string.share_to)));
                logUXEvent(FIRE_SHARE_BAR + FIRE_WHAT_IF_SUFFIX);
                return true;
            case R.id.action_go_xkcd:
                Intent browserIntent = new Intent(Intent.ACTION_VIEW, Uri.parse("https://whatif.xkcd.com/" + id));
                startActivity(browserIntent);
                logUXEvent(FIRE_GO_WHAT_IF_MENU);
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    public void updateFab() {
        if (webView.distanceToEnd() < 200) {
            scrolledToTheEnd(true);
        } else if (webView.distanceToEnd() > 250) {
            scrolledToTheEnd(false);
        }
    }

    private void scrolledToTheEnd(boolean isTheEnd) {
        if (parentFragment != null) {
            if (!parentFragment.fab.isShown() && isTheEnd) {
                parentFragment.showOrHideFabWithInfo(true);
            } else if (parentFragment.fab.isShown() && !isTheEnd) {
                parentFragment.showOrHideFabWithInfo(false);
            }
        }
    }

    @Override
    public void onImgLongClick(String title) {
        compositeDisposable.add(Observable.just(title)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::showSimpleInfoDialog,
                        e -> Timber.e("long click error")));
        if (this.getView() != null) {
            this.getView().performHapticFeedback(LONG_PRESS, FLAG_IGNORE_GLOBAL_SETTING);
        }
        logUXEvent(FIRE_WHAT_IF_IMG_LONG);
    }

    @Override
    public void onRefClick(String content) {
        compositeDisposable.add(Observable.just(content)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::showSimpleInfoDialog,
                        e -> Timber.e(e, "ref click error")));
        if (this.getView() != null) {
            this.getView().performHapticFeedback(CONTEXT_CLICK, FLAG_IGNORE_GLOBAL_SETTING);
        }
        logUXEvent(FIRE_WHAT_IF_REF);
    }

    private void showSimpleInfoDialog(String content) {
        dialog = new AlertDialog.Builder(getContext()).create();
        LinearLayout view = (LinearLayout) getLayoutInflater().inflate(R.layout.dialog_explain, null);
        TextView textView = view.findViewById(R.id.tv_explain);
        Document document = Jsoup.parse(content);
        Elements elements = document.select("img.illustration");
        for (Element element : elements) {
            element.remove();
            ImageView iv = new ImageView(getContext());
            Glide.with(getContext()).load(element.absUrl("src")).fitCenter().into(iv);
            view.addView(iv);
        }
        textView.setText(Html.fromHtml(document.html()));
        textView.setMovementMethod(LinkMovementMethod.getInstance());
        dialog.setView(view);
        dialog.show();
    }

    private class WebViewScrollCallback implements WhatIfWebView.ScrollToEndCallback {

        private WeakReference<SingleWhatIfFragment> weakReference;

        WebViewScrollCallback(SingleWhatIfFragment fragment) {
            weakReference = new WeakReference<>(fragment);
        }

        @Override
        public void scrolledToTheEnd(boolean isTheEnd) {
            if (weakReference.get() != null) {
                SingleWhatIfFragment fragment = weakReference.get();
                fragment.scrolledToTheEnd(isTheEnd);
            }
        }
    }
}

File: xkcd/src/main/java/xyz/jienan/xkcd/whatif/fragment/SingleWhatIfFragment.kt
package xyz.jienan.xkcd.whatif.fragment

import android.annotation.SuppressLint
import android.content.Intent
import android.content.res.Configuration
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.text.method.LinkMovementMethod
import android.view.HapticFeedbackConstants.*
import android.view.MenuItem
import android.view.View
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.core.os.bundleOf
import androidx.core.text.HtmlCompat
import com.bumptech.glide.Glide
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import kotlinx.android.synthetic.main.fragment_what_if_single.*
import org.jsoup.Jsoup
import org.jsoup.nodes.Element
import timber.log.Timber
import xyz.jienan.xkcd.Const.*
import xyz.jienan.xkcd.R
import xyz.jienan.xkcd.base.BaseFragment
import xyz.jienan.xkcd.model.WhatIfModel
import xyz.jienan.xkcd.model.persist.SharedPrefManager
import xyz.jienan.xkcd.model.util.appendCss
import xyz.jienan.xkcd.ui.WhatIfWebView
import xyz.jienan.xkcd.ui.getUiNightModeFlag
import xyz.jienan.xkcd.whatif.interfaces.ImgInterface
import xyz.jienan.xkcd.whatif.interfaces.LatexInterface
import xyz.jienan.xkcd.whatif.interfaces.RefInterface
import java.lang.ref.WeakReference

/**
 * Created by jienanzhang on 03/03/2018.
 */

open class SingleWhatIfFragment : BaseFragment(), ImgInterface.ImgCallback, RefInterface.RefCallback {

    var ind = -1

    private var parentFragment: WhatIfMainFragment? = null

    private val latexInterface = LatexInterface()

    protected var compositeDisposable = CompositeDisposable()

    private var dialog: AlertDialog? = null

    protected var sharedPref = SharedPrefManager

    override val layoutResId = R.layout.fragment_what_if_single

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val args = arguments
        if (args != null)
            ind = args.getInt("ind", -1)
        setHasOptionsMenu(true)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {

        setHasOptionsMenu(true)

        if (ind != -1) {
            webView.loadArticle(ind)

            webView.enableJsInterfaces()

            if (getParentFragment() is WhatIfMainFragment) {
                parentFragment = getParentFragment() as WhatIfMainFragment?
            }
        }
        WhatIfModel.observeZoom()
                .subscribe({ zoom -> webView.settings.textZoom = zoom!! },
                        { e -> Timber.e(e, "observing zoom error") })
                .let { compositeDisposable.add(it) }
    }

    override fun onDestroyView() {
        dialog?.dismiss()
        compositeDisposable.dispose()
        Timber.d("OkHttp: clear $ind")
        webView.clearTasks()
        super.onDestroyView()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (parentFragment !is WhatIfMainFragment || (parentFragment as WhatIfMainFragment).currentIndex != ind) {
            return false
        }
        when (item.itemId) {
            R.id.action_share -> {
                Intent(Intent.ACTION_SEND).apply {
                    putExtra(Intent.EXTRA_TEXT, getString(R.string.share_text_what_if, "https://whatif.xkcd.com/$ind"))
                    type = "text/plain"
                    startActivity(Intent.createChooser(this, resources.getText(R.string.share_to_what_if)))
                }
                logUXEvent(FIRE_SHARE_BAR + FIRE_WHAT_IF_SUFFIX)
            }
            R.id.action_go_xkcd -> {
                startActivity(Intent(Intent.ACTION_VIEW, Uri.parse("https://whatif.xkcd.com/$ind")))
                logUXEvent(FIRE_GO_WHAT_IF_MENU)
            }
        }
        return true
    }

    fun updateFab() {
        if (webView.distanceToEnd() < 200) {
            scrolledToTheEnd(true)
        } else if (webView.distanceToEnd() > 250) {
            scrolledToTheEnd(false)
        }
    }

    private fun scrolledToTheEnd(isTheEnd: Boolean) {
        if (parentFragment != null) {
            if (!parentFragment!!.isFabShowing && isTheEnd) {
                parentFragment!!.showOrHideFabWithInfo(true)
            } else if (parentFragment!!.isFabShowing && !isTheEnd) {
                parentFragment!!.showOrHideFabWithInfo(false)
            }
        }
    }

    override fun onImgLongClick(title: String) {
        Observable.just(title)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ showSimpleInfoDialog(it) },
                        { Timber.e(it, "long click error") })
                .also { compositeDisposable.add(it) }
        view?.performHapticFeedback(LONG_PRESS)
        logUXEvent(FIRE_WHAT_IF_IMG_LONG)
    }

    override fun onRefClick(content: String) {
        Observable.just(content)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ showSimpleInfoDialog(it) },
                        { Timber.e(it, "ref click error") })
                .also { compositeDisposable.add(it) }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            view?.performHapticFeedback(CONTEXT_CLICK)
        }
        logUXEvent(FIRE_WHAT_IF_REF)
    }

    private fun showSimpleInfoDialog(content: String) {
        dialog = AlertDialog.Builder(requireContext()).create()
        @SuppressLint("InflateParams")
        val view = layoutInflater.inflate(R.layout.dialog_explain, null) as LinearLayout
        val document = Jsoup.parse(content)
        document.convertImgToView().map { view.addView(it) }

        view.findViewById<TextView>(R.id.tvExplain).apply {
            text = HtmlCompat.fromHtml(document.html(), HtmlCompat.FROM_HTML_MODE_LEGACY)
            movementMethod = LinkMovementMethod.getInstance()
        }
        dialog!!.setView(view)
        dialog!!.show()
    }

    private fun Element.convertImgToView() =
            select("img.illustration")
                    .map { element ->
                        element.remove()
                        element.absUrl("src")
                    }.map { url ->
                        val imgView = ImageView(context)
                        Glide.with(context).load(url).fitCenter().into(imgView)
                        imgView
                    }

    private inner class WebViewScrollCallback internal constructor(fragment: SingleWhatIfFragment)
        : WhatIfWebView.ScrollToEndCallback {

        private val weakReference: WeakReference<SingleWhatIfFragment> = WeakReference(fragment)

        override fun scrolledToTheEnd(isTheEnd: Boolean) {
            val fragment = weakReference.get() ?: return
            fragment.scrolledToTheEnd(isTheEnd)
        }
    }

    private fun WhatIfWebView.loadArticle(index: Int) {
        WhatIfModel.loadArticle(index.toLong())
                .doOnSuccess { WhatIfModel.push(it) }
                .map {
                    if (context?.getUiNightModeFlag() == Configuration.UI_MODE_NIGHT_YES) {
                        val doc = Jsoup.parse(it.content)
                        doc.head()!!.appendCss("night_style.css")
                        it.content = doc.html()
                    }
                    it
                }
                .subscribe({
                    loadDataWithBaseURL("file:///android_asset/",
                            it.content?.replace("\\$".toRegex(), "&#36;"),
                            "text/html",
                            "UTF-8",
                            null)
                }, { Timber.e(it) })
                .let { compositeDisposable.add(it) }
    }

    @SuppressLint("AddJavascriptInterface")
    private fun WhatIfWebView.enableJsInterfaces() {
        setCallback(WebViewScrollCallback(this@SingleWhatIfFragment))
        setLatexScrollInterface(latexInterface)
        addJavascriptInterface(latexInterface, "AndroidLatex")
        addJavascriptInterface(ImgInterface(this@SingleWhatIfFragment), "AndroidImg")
        addJavascriptInterface(RefInterface(this@SingleWhatIfFragment), "AndroidRef")
    }

    companion object {

        fun newInstance(articleId: Int) =
                SingleWhatIfFragment().apply { arguments = bundleOf("ind" to articleId) }
    }
}